src/Farmer/Arm/Dns.fs

```diff
@@ -0,0 +1,151 @@
+[<AutoOpen>]
+module Farmer.Arm.Dns
+
+open Farmer
+open Farmer.CoreTypes
+open Farmer.Dns
+
+let zones = ResourceType ("Microsoft.Network/dnsZones", "2018-05-01")
+let aRecord = ResourceType ("Microsoft.Network/dnsZones/A", "2018-05-01")
+let aaaaRecord = ResourceType ("Microsoft.Network/dnsZones/AAAA", "2018-05-01")
+let cnameRecord = ResourceType ("Microsoft.Network/dnsZones/CNAME", "2018-05-01")
+let txtRecord = ResourceType ("Microsoft.Network/dnsZones/TXT", "2018-05-01")
+let mxRecord = ResourceType ("Microsoft.Network/dnsZones/MX", "2018-05-01")
+let nsRecord = ResourceType ("Microsoft.Network/dnsZones/NS", "2018-05-01")
+let soaRecord = ResourceType ("Microsoft.Network/dnszones/SOA", "2018-05-01")
+let srvRecord = ResourceType ("Microsoft.Network/dnsZones/SRV", "2018-05-01")
+let ptrRecord = ResourceType ("Microsoft.Network/dnsZones/PTR", "2018-05-01")
+
+type DnsRecordType with
+    member this.ResourceType =
+        match this with
+        | CName -> cnameRecord
+        | A -> aRecord
+        | AAAA -> aaaaRecord
+        | NS -> nsRecord
+        | PTR -> ptrRecord
+        | TXT -> txtRecord
+        | MX -> mxRecord
+
+type DnsZone =
+    { Name : ResourceName
+      Properties : {| ZoneType : string |} }
+
+    interface IArmResource with
+        member this.ResourceName = this.Name
+        member this.JsonModel =
+            {| zones.Create(this.Name, Location.Global) with
+                properties = {| zoneType = this.Properties.ZoneType |}
+            |} :> _
+
+module DnsRecords =
+    type CNameDnsRecord =
+        { Name : ResourceName
+          Zone : ResourceName
+          Type : DnsRecordType
+          TTL : int
+          TargetResource : ResourceName option
+          CNameRecord : string option }
+        interface IArmResource with
+            member this.ResourceName = this.Name
+            member this.JsonModel =
+                {| this.Type.ResourceType.Create(this.Zone + this.Name, dependsOn = [ this.Zone ]) with
+                    properties =
+                        {| TTL = this.TTL
+                           targetResource = {| id = this.TargetResource |}
```

Comment:
Use `Option.toObj` to correctly serialize an `Option` value (`TargetResource`).

Reply:
That gives me the following error: `The type 'ResourceName' does not have 'null' as a proper value`. Not sure what to do.

Reply:
Let me have a bash on the PR branch

Reply:
Ok, so yes, that's right :-) That's because we should also be getting the ResourceName's `Value` out i.e. the string representation. This can then safely be optioned away:

```fsharp
targetResource = {| id = this.TargetResource |> Option.map(fun r -> r.Value) |> Option.toObj |}
```

**However**, this still leaves us with a potential object with nothing in it (I'm assuming that `targetResource` is indeed optional in the ARM template spec)! So if `targetResource` is None, you'll end up with simply

```json
"targetResource": {}
```

in the ARM template. Another way around this is to do the match at the level above e.g.

```fsharp
       targetResource =
        match this.TargetResource with
        | Some targetResource -> box {| id = targetResource.Value |}
        | None -> null
```

This will completely omit the `targetResource` field and object if the value in F# is None.

Reply:
According to the ARM template everything is optional hehe.

I've checked with our an existing DNS Zone in Azure to make sure, `"targetResource": {}` is valid when it's empty.

I guess the following would also work.
```fsharp
this.TargetResource
|> Option.map (fun x -> {| id = x.Value |}
|> Option.toObj
```

What solution shall we pick?